redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt

```diff
@@ -15,118 +15,156 @@
 
-      holder.widgetContentBinding?.close()
-      holder.widgetContentBinding = itemContentSource.bindWhenReady(
-        view = holder.treehouseWidgetView,
-        app = treehouseApp,
+    override suspend fun load(
+      params: LoadParams<Int>,
+    ): LoadResult<Int, Content> {
+      val key = params.key ?: 0
+      val count = intervals.sumOf(LazyListIntervalContent::count)
+      val limit = when (params) {
+        is LoadParams.Prepend<*> -> minOf(key, params.loadSize)
+        else -> params.loadSize
+      }.coerceAtMost(count)
+      val offset = when (params) {
+        is LoadParams.Prepend<*> -> maxOf(0, key - params.loadSize)
+        is LoadParams.Append<*> -> key
+        is LoadParams.Refresh<*> -> if (key >= count) maxOf(0, count - params.loadSize) else key
+      }
+      val nextPosToLoad = offset + limit
+      val loadResult = LoadResult.Page(
+        data = List(limit) { index ->
+          val itemContentSource = TreehouseContentSource<A> {
+            var interval = IndexedValue(index + offset, intervals.first())
+            for (nextInterval in intervals.drop(1)) {
+              if (interval.index < interval.value.count) break
+              interval = IndexedValue(interval.index - interval.value.count, nextInterval)
+            }
+            interval.value.itemProvider.get(interval.index)
+          }
+          treehouseApp.createContent(itemContentSource).apply {
+            // TODO Pass in actual HostConfiguration
+            preload(HostConfiguration(darkMode = true))
+            awaitContent()
+          }
+        },
+        prevKey = offset.takeIf { it > 0 && limit > 0 },
+        nextKey = nextPosToLoad.takeIf { limit > 0 && it < count },
+        itemsBefore = offset,
+        itemsAfter = maxOf(0, count - nextPosToLoad),
       )
+      return if (invalid) LoadResult.Invalid() else loadResult
     }
+
+    override fun getRefreshKey(state: PagingState<Int, Content>) =
+      state.anchorPosition?.let { maxOf(0, it - (state.config.initialLoadSize / 2)) }
   }
 
-  private class ViewHolder(
-    container: FrameLayout,
-    widgetSystem: WidgetSystem,
-  ) : RecyclerView.ViewHolder(container) {
-    var widgetContentBinding: Closeable? = null
-    val treehouseWidgetView = TreehouseWidgetView(container.context, widgetSystem)
-      .apply {
+  private class LazyContentItemListAdapter(
+    private val widgetSystem: WidgetSystem,
+  ) : PagingDataAdapter<Content, ViewHolder>(ContentDiffCallback) {
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder = ViewHolder(
+      TreehouseWidgetView(parent.context, widgetSystem).apply {
         layoutParams = FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
           gravity = Gravity.CENTER_HORIZONTAL
         }
+      },
+    )
 
-        container.addView(this)
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+      getItem(position)!!.apply {
+        unbind()
```

Comment:
Uh oh, I think this might destroy what we’ve preloaded. Fix would be to move unbind() somewhere else . . .

Reply:
Doh! Moved the `unbind` call to `onViewRecycled`.